atc/configvalidate/validate.go

```diff
@@ -490,3 +496,62 @@ func validateDisplay(c atc.Config) ([]atc.ConfigWarning, error) {
 
 	return warnings, nil
 }
+
+func detectCycle(j atc.JobConfig, visited map[string]int, initialJobConfig atc.Config) (bool, string) {
+	const (
+		nonVisited = 0
+		semiVisited = 1
+		alreadyVisited = 2
+	)
+	visited[j.Name] = semiVisited
+	state := false
+	cycleErrorJob := ""
+	_ = j.StepConfig().Visit(atc.StepRecursor{
+		OnGet: func(step *atc.GetStep) error {
+			for _, nextJobName := range step.Passed {
+				nextJob := findJobByName(nextJobName, initialJobConfig.Jobs)
+				if visited[nextJobName] == semiVisited {
+					state = true
+					cycleErrorJob = nextJobName
+					return nil
```

Comment:
Not a big deal, but WDYT about changing the signature of `detectCycle` to just return an `error`. If we detect a cycle here, we can just return the error we currently construct here: https://github.com/concourse/concourse/blob/c948550fecc2a8df7fd178364aed7d145b15cc80/atc/configvalidate/validate.go#L554

Then, we can simply return the `error` that's returned by `j.StepConfig.Visit(...)`, and don't need to keep track of `state` or `cycleErrorJob`, simplifying the control flow a bit. Plus, it wasn't totally obvious to me what `state` referred to initially - but returning an `error` makes it more obvious that this is where the cycle occurs